File: src/main/java/ca/frar/jjjrmi/translator/TranslatorResult.java
package ca.frar.jjjrmi.translator;

import ca.frar.jjjrmi.annotations.JSParam;
import ca.frar.jjjrmi.exceptions.DecoderException;
import ca.frar.jjjrmi.exceptions.EncoderException;
import ca.frar.jjjrmi.exceptions.JJJRMIException;
import ca.frar.jjjrmi.exceptions.UnknownReferenceException;
import ca.frar.jjjrmi.exceptions.RootException;
import ca.frar.jjjrmi.utility.JJJOptionsHandler;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import org.json.JSONObject;

/**
 * An array (wrapped) of encoded objects. The first object is the target object,
 * all other object are object that have not been previously translated.
 *
 * @author Ed Armstrong
 */
public class TranslatorResult {
    private final Translator translator;
    private JSONObject json;
    private Object root;

    /**
     * For one off encodings.
     * @param object
     */
    public static TranslatorResult fromObject(Object object) throws JJJRMIException {
        return new Translator().encode(object);
    }

    /**
     * For creating one off decodings.
     * @param string
     */
    public static TranslatorResult fromString(String string) throws JJJRMIException {
        return new Translator().decode(string);
    }

    TranslatorResult(Translator translator) {
        this.translator = translator;
    }

    TranslatorResult encodeFromObject(Object object) throws JJJRMIException {
        if (object == null) throw new RootException();
        this.json = new JSONObject();
        this.json.put(Constants.NewObjects, new JSONObject());

        if (this.translator.hasReferredObject(object)) {
            this.setRoot(this.translator.getReference(object));
        } else if (HandlerFactory.getInstance().hasHandler(object.getClass())) {
            encodeHandled(object);
        } else {
            encodeUnhandled(object);
        }

        return this;
    }

    private void encodeHandled(Object object) throws JJJRMIException {
        try {
            Class<? extends AHandler<?>> handlerClass = HandlerFactory.getInstance().getHandler(object.getClass());
            AHandler<?> handler = handlerClass.getConstructor(TranslatorResult.class).newInstance(this);
            EncodedObject encodedObject = handler.doEncode(object);
            this.put(encodedObject);
            this.setRoot(this.translator.getReference(object));
        } catch (NoSuchMethodException | SecurityException | InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException ex) {
            throw new EncoderException(ex, object);
        }
    }

    private void encodeUnhandled(Object object) throws JJJRMIException {
        try {
            EncodedObject encodedObject = new EncodedObject(object, this, new JJJOptionsHandler(object).retain());
            this.put(encodedObject);
            this.setRoot(this.translator.getReference(object));
            encodedObject.encode();
        } catch (IllegalArgumentException | IllegalAccessException ex) {
            throw new EncoderException(ex, object);
        }
    }

    TranslatorResult decodeFromString(String source) throws DecoderException {
        this.json = new JSONObject(source);
        ArrayList<ObjectDecoder> list = new ArrayList<>();
        JSONObject newObjects = this.json.getJSONObject(Constants.NewObjects);

        for (String key : newObjects.keySet()) {
            if (this.translator.hasReference(key)) continue;
            JSONObject jsonObject = newObjects.getJSONObject(key);
            list.add(new ObjectDecoder(this, jsonObject));
        }
        for (ObjectDecoder decoder : list) {
            decoder.makeReady();
        }
        for (ObjectDecoder decoder : list) {
            decoder.decode();
        }
        for (ObjectDecoder decoder : list) {
            try {
                decoder.afterDecode();
            } catch (InvocationTargetException | IllegalAccessException e) {
                throw new DecoderException(e);
            }
        }

        return this;
    }

    /**
     * @return the translator
     */
    Translator getTranslator() {
        return translator;
    }

    void finalizeRoot() throws UnknownReferenceException {
        String key = this.json.getString(Constants.RootObject);
        this.root = this.translator.getReferredObject(key);
    }

    void setRoot(String rootKey) {
        if (rootKey == null) throw new NullPointerException();
        this.json.put(Constants.RootObject, rootKey);
    }

    public Object getRoot() throws UnknownReferenceException {
        return this.root;
    }

    void put(EncodedObject encodedObject) {
        String key = encodedObject.getKey();
        JSONObject toJSON = encodedObject.toJSON();
        this.json.getJSONObject(Constants.NewObjects).put(key, toJSON);
    }

    /**
     * Returns the number of new objects in this result.
     *
     * @return the number of new objects.
     */
    public int newObjectCount() {
        return this.json.getJSONObject(Constants.NewObjects).keySet().size();
    }

    public JSONObject toJSON() {
        return this.json;
    }

    public String toString() {
        return this.json.toString();
    }

    @JSParam(name = "indentFactor", init = "0")
    public String toString(int indentFactor) {
        return this.json.toString(indentFactor);
    }
}

File: src/main/js/translator/TranslatorResult.js
"use strict";
const Constants = require("./Constants");
const EncodedObject = require("./Encoder").EncodedObject;
const ObjectDecoder = require("./ObjectDecoder");
const LOGGER = require("./Logger");

class TranslatorResult {
    constructor(translator) {
        this.translator = translator;
    }
    decodeFromString(source) {
        this.json = JSON.parse(source);
        let list = [];
        let newObjects = this.json[Constants.NewObjects];

        for (let key in newObjects) {
            if (this.translator.hasReference(key)) continue;
            let jsonObject = newObjects[key];
            list.push(new ObjectDecoder(this, jsonObject, this.translator));
        }
        for (let decoder of list) {
            decoder.makeReady();
        }
        for (let decoder of list) {
            decoder.decode();
        }
        for (let decoder of list) {
            decoder.afterDecode();
        }
        return this;
    }
    encodeFromObject(object) {
        LOGGER.log("translator", `encodeFromObject(${object.constructor.name})`);
        let handlerRegistry = this.translator.handlerRegistry;

        this.json = {};
        this.json[Constants.NewObjects] = {};
        if (this.translator.hasReferredObject(object)){
            LOGGER.log("translator", `hasReferredObject : ${this.translator.getReference(object)}`);
            this.setRoot(this.translator.getReference(object));
        }
        else if (object.constructor.__getClass && handlerRegistry.hasClass(object.constructor.__getClass())){
            LOGGER.log("translator", `encodeHandled`);
            this.encodeHandled(object);
        }
        else{
            LOGGER.log("translator", `encodeUnhandled`);
            this.encodeUnhandled(object);
        }
        
        return this;
    }
    encodeHandled(object) {
        let handlerRegistry = this.translator.handlerRegistry;
        let handlerClass = handlerRegistry.getClass(object.constructor.__getClass());
        let handler = new handlerClass(this);
        let encodedObject = handler.doEncode(object);
        this.put(encodedObject);
        this.setRoot(this.translator.getReference(object));
    }
    encodeUnhandled(object) {
        let encodedObject = new EncodedObject(object, this, object.constructor.__isRetained());
        this.put(encodedObject);
        this.setRoot(this.translator.getReference(object));
        encodedObject.encode();
    }
    getRoot() {
        return this.root;
    }
    getTranslator() {
        return this.translator;
    }
    newObjectCount() {
        return this.json[Constants.NewObjects].keySet().size();
    }
    put(encodedObject) {
        let key = encodedObject.getKey();
        let toJSON = encodedObject.toJSON();
        this.json[Constants.NewObjects][key] = toJSON;
    }
    setRoot(rootKey) {
        this.json[Constants.RootObject] = rootKey;
    }
    finalizeRoot(){
        let key = this.json[Constants.RootObject];
        this.root = this.translator.getReferredObject(key);
    }    
    toString(indentFactor = 0) {
        return JSON.stringify(this.json, null, indentFactor);
    }
    toJSON(){
        return this.json;
    }
}

if (typeof module !== "undefined") module.exports = TranslatorResult;